Feature/issue-27/guided-color-prompt

KoatKoetl opened the pull request:
Closes #27 
## Description
I created the guided color form for users. The PR include the component itself + addditional installation of Shadcn/UI components that I used to build the form. The component that I work on will look like this in the end:

On PC
![image](https://github.com/chingu-voyages/v49-tier2-team-15/assets/68298104/932a29d5-ecc2-482b-8643-fc6d65220fd8)

On mobile 
![image](https://github.com/chingu-voyages/v49-tier2-team-15/assets/68298104/a07e6948-6268-45fe-b9ca-1a992fb7aa5e)


### Work
The work isn't done 100%, I have to style all and do some logic for the API. 
I build the logic to check the inputs, for some of them I check to have only one word no spaces, for others I have limit to 255 characters. Also I implemented keyword select method which contain 5 default values + added a field to input your custom keyword.

Keywords input look like this
![FormGIF](https://github.com/chingu-voyages/v49-tier2-team-15/assets/68298104/2477484b-fd5a-492f-a3b5-65a75b4304b7)

### Need Help/Advice
I'm just submitting a request to get advice on the work I've done.
1. I have created a custom form submit function that sends input to the openAI API. In response I can get colors inside my component, but those colors will be used in the DisplayColors component for example. 
So my main problem is how can I make it so that the colors can be used in other components. If anyone can give me some advice on this, it would be great.
2. Also, if you have time, give me some tips on code structure, because I'm not sure it's written well enough, and I myself don't see any ideas on how to change the code even better.

KoatKoetl committed:
Add - Shadcn components to build the form

KoatKoetl committed:
Create a UserGuideComponent with a form that has different input validations

KoatKoetl committed:
Add additional check for the audience input + fix some css

KoatKoetl committed:
Install - Select Shadcn component

KoatKoetl committed:
Fix - some logic on handleChange + add keywords select

KoatKoetl committed:
Fix - some of the logic for inputs + create a POST to the API on form submit

feliperdamaceno commented on src/components/userGuidedForm.tsx:
The component owns too much helper fuctions in a single file, most of them can me moved to a separate file in the utils folder.

feliperdamaceno commented on src/components/userGuidedForm.tsx:
```diff
@@ -0,0 +1,263 @@
+} from '@/components/ui/select';
+import { createGuidedColorPrompt } from '@/helpers/generators';
+
+interface userGuidedFormProps {
```
Victoria will work in a theme hook, which we'll use to get the initial color from whatever the user have selected in the Color Picker Nicola is building. Essentially we'll do something like this instead:

```javascript
const { initialColor } = useContext(ColorContext)
```

feliperdamaceno commented on src/components/userGuidedForm.tsx:
The component owns too much helper fuctions in a single file, most of them can me moved to a separate file in the utils folder.

Also, is standard for component files to be CamelCase, I personally would use GuidedForm.tsx only.

feliperdamaceno commented on src/components/userGuidedForm.tsx:
```diff
@@ -0,0 +1,263 @@
+  const [keywordsError, setKeywordsError] = useState<boolean>(false);
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { id, value } = event.target;
```
It is not completely wrong, but what you need here instead is the **name** property intead of Id. Id's are CSS related property and for form validation, **name** is most used.

feliperdamaceno commented on src/components/userGuidedForm.tsx:
```diff
@@ -0,0 +1,263 @@
+        setKeywords(value);
+        break;
+      default:
+        break;
+    }
+  };
+
+  const countOneWord = (event: React.ChangeEvent<HTMLInputElement>) => {
+    return event.target.value.split(' ').length === 1;
+  };
+
+  const countChars = (event: React.ChangeEvent<HTMLInputElement>) => {
+    return event.target.value.length <= 255;
+  };
+
+  const handleSelectChange = (value: string) => {
+    if (!selectedKeywords.includes(value) && selectedKeywords.length < 5) {
+      setSelectedKeywords([...selectedKeywords, value]);
+    }
+  };
+
+  const handleAddCustomKeyword = () => {
+    if (
+      keyword &&
+      !selectedKeywords.includes(keyword) &&
+      selectedKeywords.length < 5
+    ) {
+      setSelectedKeywords([...selectedKeywords, keyword]);
+      setKeywords('');
+    } else if (selectedKeywords.length >= 5) {
+      setKeywordsError(true);
+    } else {
+      setKeywordsError(false);
+    }
+  };
+
+  const handleDeleteKeyword = (keyword: string) => {
+    setSelectedKeywords(selectedKeywords.filter((word) => word !== keyword));
+    setKeywordsError(false);
+  };
```
When we have such a complex state and validation, is better to use something like useReducer instead of useState, so that we can keep the business logic separated from the presentation. The main problem is that all those helper functions are inside the form component, which will cause all of them to be recreated every time users type in any of the fields, which can cause serious hit in performance, and also it makes the components too much overwelming and hard to manage.

You can check more here: https://joshua777.hashnode.dev/how-to-use-the-usereducer-hook-for-simple-form-validation-in-react

Another alternative is using validation libraries like Zod, so that you keep your component dry, or maybe libraries like react-hook-form.

feliperdamaceno commented on src/components/userGuidedForm.tsx:
```diff
@@ -0,0 +1,263 @@
+                <Label htmlFor="purpose">I need my color scheme for...</Label>
+                <Input
+                  id="purpose"
+                  placeholder="Website, poster... (max: 255 characters)"
```
We don't actually need to limit to 255 characters in this scenario, is actually better for the AI context having a good prompt. However, in cases that we need to include that, thinking in a UX perspective, is always better to not include things like (max: 255 characters) in the placeholder, let the user make that mistake, **then** you show the errors.

Think like that, why should the user be blamed by an error that never happend? 🙂

feliperdamaceno commented on src/components/userGuidedForm.tsx:
```diff
@@ -0,0 +1,263 @@
+                    value={keyword}
+                    onChange={handleChange}
+                  />
+                  <Select onValueChange={handleSelectChange}>
+                    <SelectTrigger className="w-12" />
+                    <SelectContent>
+                      {predefinedKeywords.map((keyword, index) => (
+                        <SelectItem key={index} value={keyword}>
+                          {keyword}
+                        </SelectItem>
+                      ))}
+                    </SelectContent>
+                  </Select>
```
This one is a big tricky, it is not being used in a UX focused way I guess. Using a select to choose the keywords is a bit confusing IMO. I think would be better to have 5 pill components where the user can click to add into the list, but TBH I think we don't actually need those default values. For this scenario, I think would be better to let the user add whatever they prefer.

Also, there are a few missing parts on this component like SelectGroup and SelectValue where you can add a placeholder for a better UX. Is a bit hard for the users to know what this select does straightaway, maybe include a placeholder with as message?

```javascript
<SelectValue placeholder="Select a keywords" />
```

But again, not sure we need that though.

feliperdamaceno reviewed (changes requested):
If you want to improve the component based my feedback feel free to do so, but I don't think that would be 100% required.

This is not a real production application, so changing or not won't affect the app so much, but I'll leave that decision to you ✌🏻🐙

KoatKoetl committed:
Rename file

KoatKoetl committed:
Move helper function to separate files

KoatKoetl committed:
Move all function inside a related folder

KoatKoetl committed:
Separat in files helper functions + remove unused one

KoatKoetl commented:
Hi. I tried to do my best to change the code using your feedback. 
1. I've changed the file name
2. Removed the props from the component and wait until the hook is done.
3. Removed the select for keywords. Now users can type whatever they want.
4. Removed the 255 character limit. I just wanted to limit the user from writing a lot of text, but I agree that without the limit the AI will give more accurate answers.
5. Moved all helpers functions outside the component.
6. I also tried to do the best I could with UseReducer. The form works exactly the same way.
Now I'm going to start styling the component.

KoatKoetl committed:
Fix Component name

KoatKoetl committed:
Style the form

VriaA committed:
fix: remove redundant error messages

- Removed error messages from the usage, audience, and mood input fields.
- Made keywords optional.

VriaA committed:
remove guided form from App.tsx

VriaA committed:
chore: format code with Prettier
